### ISSUE TITLE
Smart select does not select TypeScript object type

### ISSUE THREAD

**Issue by OliverJAsh** [microsoft/TypeScript#39618]
- VSCode Version: latest
- OS Version: macOS latest

Steps to Reproduce:

1. Paste this into a TypeScript file
   ```ts
   type Foo = {
     /** comment */
     status: number;
   };
   ```
2. Place cursor just before `status`
3. Run command "expand selection" 4 times

Expected:  only the object should be selected:

![image](https://user-images.githubusercontent.com/921609/87570614-cfc47280-c6c0-11ea-990b-e69ad7a4a818.png)

Actual: the whole type is selected:

![image](https://user-images.githubusercontent.com/921609/87570647-dbb03480-c6c0-11ea-963d-6e86bb8b247d.png)

![gif](https://user-images.githubusercontent.com/921609/87570997-4fead800-c6c1-11ea-8261-f68934f6f281.gif)

Does this issue occur when all extensions are disabled?: Yes

**Comment by McZenith** [comment]
Hello! I'm kinda new... Can I grab this?

**Comment by heanzyzabala** [comment]
Hi, @McZenith are you currently working on this?

**Comment by McZenith** [comment]
I was just asking if I can...

**Comment by ShahzaibP** [comment]
Hello, looking for my first contribution. Can anyone confirm if this is assigned to someone? If not, I would like to pitch in. Thanks.

**Comment by saigeethakandepallicherukuru** [comment]
Hello, This my first time attending vGHC and I am looking for my first open source contribution. Can anyone let me know if I can work on this issue? Thanks!

**Comment by RyanCavanaugh** [comment]
@saigeethakandepallicherukuru please go ahead!

Everyone else: Sorry to have not responded before, but No Assignee (up at the top on the right) means no TS team member is working on the bug and it's up for grabs. Unless there's already an open PR, it's a pretty safe bet that no one else is working on any particular bug.

**Comment by denismaxim0v** [comment]
I'd love to work on that

**Comment by marianayap** [comment]
Hello! @OliverJAsh is this issue take it or could I help?

**Comment by andrewbranch** [comment]
I’m going to use this one as an example for a workshop—we might fix it during the workshop, but if not, I’ll unassign myself. At that point, anyone who wants to is free to take it and put up a PR.

**Comment by andrewbranch** [comment]
Actually, this seems to be fixed already.

**Comment by OliverJAsh** [comment]
Unfortunately I don't think it is fixed, it's just that the behaviour has changed slightly. The object is still not selected—the whole file is selected instead.

New steps to reproduce:

1. Paste this into a TypeScript file
   ```ts
   const x = 1;

   type Foo = {
     /** comment */
     status: number;
   };
   ```
2. Place cursor just before `status`
3. Run command "expand selection" 4 times

Expected: only the object should be selected:

![image](https://user-images.githubusercontent.com/921609/102486737-6676d000-4061-11eb-991a-326acf702abf.png)

Actual: the whole file is selected:

![image](https://user-images.githubusercontent.com/921609/102486726-5fe85880-4061-11eb-9996-168c8e509e05.png)

@andrewbranch Can we re-open this?

**Comment by andrewbranch** [comment]
Hmm, it seems like the comment is the thing throwing it off, but weirdly it works perfectly in the playground. Given that, I suspect that another one of VS Code's selection range providers is interfering.

**Comment by andrewbranch** [comment]
This issue is again up for grabs, but it’s possible the result is that you discover this is a VS Code bug, not a TypeScript bug.

**Comment by bhupinderr8** [comment]
Tried the expand selection 4 times with a js file

![Screenshot from 2021-10-06 00-43-13](https://user-images.githubusercontent.com/28729320/136087974-0b5370be-b588-44f6-af7c-3ffb1766b730.png)

Without second asterisk:
![Screenshot from 2021-10-06 00-45-01](https://user-images.githubusercontent.com/28729320/136088275-2d70c4ec-3761-4dae-9fe6-106d6e99a6ae.png)

The issue is not specific to ts files, looks like some VScode bug.

**Comment by babakks** [comment]
I tried to trace the problem on the VS Code side and found some cues that the problem lies within the TS language server.

## Problematic case

This is the input code block that I used:

![test ts](https://user-images.githubusercontent.com/36728931/176965237-114c06da-9d58-4a58-95f8-b19eca9e6ba8.png)

I put the caret at the beginning of the `status` field (position 5:3) and invoked the `Expand Selection` command. This is the (flattened) tree of ranges and their parents returned from the TypeScript language server (for any range its parent sits at the next line):

```text
5:3 >> 5:9
4:3 >> 5:9
5:3 >> 5:18 (*)
4:3 >> 5:18
3:13 >> 6:1
3:12 >> 6:2
3:1 >> 6:3
1:1 >> 8:15
```

VS Code editor expects parents to include their child, but this is broken at the third line (here indicated by the `*`). VS Code's behavior is to drop all ranges due to this inconsistency.

## Healthy case

I also tried this with the healthy code sample below:

![code ts-ok](https://user-images.githubusercontent.com/36728931/176965982-bf03f6a2-f42d-414b-8a87-2b60416b3687.png)

Again, I put the caret at the beginning of the `status` field (position 5:3) and invoked the `Expand Selection` command. This is the array of resulting ranges, which satisfies the parent-child criterion:

```text
4:3 >> 4:9
4:3 >> 4:18
3:13 >> 5:1
3:12 >> 5:2
3:1 >> 5:3
1:1 >> 7:15
```

@andrewbranch If the fix should be done on TS, I'd appreciate any hints to begin with since I'm new to the codebase.

**Comment by andrewbranch** [comment]
Yep, that confirms the problem is ours. All that code is [here](https://github.com/microsoft/TypeScript/blob/main/src/services/smartSelection.ts). If you add a new test copying the format of [this one](https://github.com/microsoft/TypeScript/blob/main/tests/cases/fourslash/smartSelection_objectTypes.ts), you should be able to debug that code and see what’s going on:

- https://github.com/microsoft/TypeScript/blob/main/CONTRIBUTING.md#debugging-the-tests
- https://blog.andrewbran.ch/debugging-the-type-script-codebase/

**Comment by babakks** [comment]
@andrewbranch I've submitted the PR. Could you please check it?

Also, I've found another similar behavior in smart selection on function parameters. Since I'm not sure if that should be fixed with the current PR I've filed a separate issue.
